File: daysPastEpoch.c
// daysPastEpoch.c
// determine days past an epoch, the first day in year 1900

// This code is a port of daysPastEpoch.lua to c

#include <assert.h>
#include <math.h>
#include <stdio.h>

#include "daysPastEpoch.h"

static unsigned isLeapYear(unsigned year)
{
  if ((year % 100) == 0) {
    if ((year % 400) == 0)
      return 0;
    else
      return 1;
  }
  else if ((year % 4) == 0)
    return 1;
  else
    return 0;
}

static void check(double expected, double date)
{
  double actual = daysPastEpoch(date);
  if (expected == actual) return;
  fprintf(stderr, "unit test of daysPastEpoch failed\n");
  fprintf(stderr, "expected = %f\n", expected);
  fprintf(stderr, "actual   = %f\n", actual);
  assert(expected == actual);  // to get the stack walkback
}

static void unitTest()
{
  check(  0, 19000101);
  check(  1, 19000102);
  check( 31, 19000201);
  check( 60, 19000301);
  check(365, 19001231);
  check(366, 19010101);

}

double daysPastEpoch(double date)
{
  static unsigned unitTested = 0;
  if (!unitTested) {
    unitTest();
    unitTested = 1;
  }

  assert(date > 19000000);
  double year = floor(date / 10000);
  double month = floor(date / 100 - year * 100);
  double day   = floor(date - year * 10000 - month * 100);

  const double firstYear = 1900;
  double yearDays = 0;
  for (unsigned yearNumber = firstYear; yearNumber < year - 1; yearNumber++) {
    yearDays += 365 + isLeapYear(yearNumber);
  }

  static double daysPerMonth[] = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
  double monthDays = 0;
  for (unsigned monthNumber = 1; monthNumber < month; monthNumber++) {
    monthDays += daysPerMonth[monthNumber];
  }

  double leapDay = 0;
  if (isLeapYear((long)year) && month > 2)
    leapDay = 1;

  return yearDays + monthDays + leapDay + day - 1;
}

File: lua/daysPastEpoch.lua
-- daysPastEpoch.lua
-- define function return number of days past an epoch (1900-01-01)
-- The old epoch date is needed because some transactions date to 1908
-- and I don't want to use negative day numbers (as uncertain what the
-- effect would be).

do
   local daysPerMonth = 
      {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
   
   local function isLeapYear(year)
      if (year % 100) == 0 then
         if (year % 400)  == 0 then
            return 0
         else
            return 1
         end
      elseif  (year % 4) == 0 then 
         return 1 
      else 
         return 0 
      end
   end

   local firstYear = 1900
   
   function daysPastEpoch(date)
      local trace = false
      local year  = math.floor(date / 10000)
      local month = math.floor(date / 100 - year * 100)
      local day   = math.floor(date - year * 10000 - month * 100)
      if trace then print(date, year, month, day) end
      
      -- new version
      assert(year >= firstYear)
      
      local yearDays = 0
      for yearNumber = firstYear, year - 1 do
         yearDays = yearDays + 365 + isLeapYear(yearNumber)
      end
      
      local monthDays = 0
      for monthNumber = 1, month - 1 do
         monthDays = monthDays + daysPerMonth[monthNumber]
      end
      
      local leapDay = 0
      if isLeapYear(year) and month > 2 then
         leapDay = 1
      end
      
      local result = yearDays + monthDays + leapDay + day - 1
      if trace then print(yearDays, monthDays, leapDay, result) end
      
      return result
   end
end